File: apps/android/shared/src/main/java/com/eatfair/shared/data/remote/NegotiationService.kt
package com.eatfair.shared.data.remote

import com.eatfair.shared.config.AppConfig
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import okio.ByteString
import org.json.JSONObject
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Negotiation Service
 * Matches iOS NegotiationService.swift for platform parity
 * Real-time price negotiation via WebSocket
 */

// Data Models
data class Negotiation(
    val id: String,
    val orderId: String? = null,
    val rideId: String? = null,
    val customerId: String,
    val driverId: String? = null,
    val suggestedPrice: Double,
    val currentPrice: Double,
    val customerOffer: Double? = null,
    val driverOffer: Double? = null,
    val platformFee: Double,
    val status: NegotiationStatus,
    val createdAt: Long,
    val expiresAt: Long
)

enum class NegotiationStatus {
    PENDING,
    CUSTOMER_OFFERED,
    DRIVER_OFFERED,
    ACCEPTED,
    REJECTED,
    EXPIRED,
    CANCELLED
}

data class NegotiationEvent(
    val type: NegotiationEventType,
    val negotiationId: String,
    val data: Map<String, Any>
)

enum class NegotiationEventType {
    PRICE_UPDATE,
    CUSTOMER_OFFER,
    DRIVER_OFFER,
    ACCEPTED,
    REJECTED,
    EXPIRED,
    ERROR
}

data class CreateNegotiationRequest(
    val orderId: String? = null,
    val rideId: String? = null,
    val customerId: String,
    val pickupLat: Double,
    val pickupLng: Double,
    val dropoffLat: Double,
    val dropoffLng: Double,
    val estimatedDistance: Double,
    val estimatedDuration: Int
)

data class CounterOfferRequest(
    val negotiationId: String,
    val offeredPrice: Double
)

sealed class NegotiationResult<out T> {
    data class Success<T>(val data: T) : NegotiationResult<T>()
    data class Error(val message: String) : NegotiationResult<Nothing>()
}

@Singleton
class NegotiationService @Inject constructor() {

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private var webSocket: WebSocket? = null

    // Use AppConfig for environment-aware URLs (production)
    private val baseUrl: String
        get() = AppConfig.API_BASE_URL.removeSuffix("/api")

    private val wsUrl: String
        get() {
            val httpUrl = AppConfig.API_BASE_URL.removeSuffix("/api")
            return httpUrl.replace("http://", "ws://").replace("https://", "wss://")
        }

    private val _connectionState = MutableStateFlow(ConnectionState.DISCONNECTED)
    val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()

    private val _negotiationEvents = MutableSharedFlow<NegotiationEvent>()
    val negotiationEvents: Flow<NegotiationEvent> = _negotiationEvents

    private val _currentNegotiation = MutableStateFlow<Negotiation?>(null)
    val currentNegotiation: StateFlow<Negotiation?> = _currentNegotiation.asStateFlow()

    enum class ConnectionState {
        DISCONNECTED,
        CONNECTING,
        CONNECTED,
        ERROR
    }

    /**
     * Create a new negotiation session
     */
    suspend fun createNegotiation(
        request: CreateNegotiationRequest,
        token: String
    ): NegotiationResult<Negotiation> {
        return try {
            val json = JSONObject().apply {
                request.orderId?.let { put("order_id", it) }
                request.rideId?.let { put("ride_id", it) }
                put("customer_id", request.customerId)
                put("pickup_lat", request.pickupLat)
                put("pickup_lng", request.pickupLng)
                put("dropoff_lat", request.dropoffLat)
                put("dropoff_lng", request.dropoffLng)
                put("estimated_distance", request.estimatedDistance)
                put("estimated_duration", request.estimatedDuration)
            }

            val requestBody = json.toString().toRequestBody("application/json".toMediaType())

            val httpRequest = Request.Builder()
                .url("$baseUrl/api/negotiations")
                .addHeader("Authorization", "Bearer $token")
                .addHeader("Content-Type", "application/json")
                .post(requestBody)
                .build()

            val response = client.newCall(httpRequest).execute()

            if (response.isSuccessful) {
                val responseBody = response.body?.string()
                val negotiation = parseNegotiation(JSONObject(responseBody ?: "{}"))
                _currentNegotiation.value = negotiation
                NegotiationResult.Success(negotiation)
            } else {
                NegotiationResult.Error("Failed to create negotiation: ${response.message}")
            }
        } catch (e: Exception) {
            NegotiationResult.Error(e.message ?: "Unknown error")
        }
    }

    /**
     * Submit customer counter-offer
     */
    suspend fun submitCustomerOffer(
        negotiationId: String,
        offeredPrice: Double,
        token: String
    ): NegotiationResult<Negotiation> {
        return try {
            val json = JSONObject().apply {
                put("offered_price", offeredPrice)
            }

            val requestBody = json.toString().toRequestBody("application/json".toMediaType())

            val httpRequest = Request.Builder()
                .url("$baseUrl/api/negotiations/$negotiationId/customer-offer")
                .addHeader("Authorization", "Bearer $token")
                .addHeader("Content-Type", "application/json")
                .post(requestBody)
                .build()

            val response = client.newCall(httpRequest).execute()

            if (response.isSuccessful) {
                val responseBody = response.body?.string()
                val negotiation = parseNegotiation(JSONObject(responseBody ?: "{}"))
                _currentNegotiation.value = negotiation
                NegotiationResult.Success(negotiation)
            } else {
                NegotiationResult.Error("Failed to submit offer: ${response.message}")
            }
        } catch (e: Exception) {
            NegotiationResult.Error(e.message ?: "Unknown error")
        }
    }

    /**
     * Submit driver counter-offer
     */
    suspend fun submitDriverOffer(
        negotiationId: String,
        offeredPrice: Double,
        token: String
    ): NegotiationResult<Negotiation> {
        return try {
            val json = JSONObject().apply {
                put("offered_price", offeredPrice)
            }

            val requestBody = json.toString().toRequestBody("application/json".toMediaType())

            val httpRequest = Request.Builder()
                .url("$baseUrl/api/negotiations/$negotiationId/driver-offer")
                .addHeader("Authorization", "Bearer $token")
                .addHeader("Content-Type", "application/json")
                .post(requestBody)
                .build()

            val response = client.newCall(httpRequest).execute()

            if (response.isSuccessful) {
                val responseBody = response.body?.string()
                val negotiation = parseNegotiation(JSONObject(responseBody ?: "{}"))
                _currentNegotiation.value = negotiation
                NegotiationResult.Success(negotiation)
            } else {
                NegotiationResult.Error("Failed to submit offer: ${response.message}")
            }
        } catch (e: Exception) {
            NegotiationResult.Error(e.message ?: "Unknown error")
        }
    }

    /**
     * Accept current negotiated price
     */
    suspend fun acceptNegotiation(
        negotiationId: String,
        token: String
    ): NegotiationResult<Negotiation> {
        return try {
            val httpRequest = Request.Builder()
                .url("$baseUrl/api/negotiations/$negotiationId/accept")
                .addHeader("Authorization", "Bearer $token")
                .addHeader("Content-Type", "application/json")
                .post("".toRequestBody(null))
                .build()

            val response = client.newCall(httpRequest).execute()

            if (response.isSuccessful) {
                val responseBody = response.body?.string()
                val negotiation = parseNegotiation(JSONObject(responseBody ?: "{}"))
                _currentNegotiation.value = negotiation
                NegotiationResult.Success(negotiation)
            } else {
                NegotiationResult.Error("Failed to accept: ${response.message}")
            }
        } catch (e: Exception) {
            NegotiationResult.Error(e.message ?: "Unknown error")
        }
    }

    /**
     * Connect to WebSocket for real-time updates
     */
    fun connectWebSocket(negotiationId: String, token: String) {
        _connectionState.value = ConnectionState.CONNECTING

        val request = Request.Builder()
            .url("$wsUrl/ws/negotiation/$negotiationId")
            .addHeader("Authorization", "Bearer $token")
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                _connectionState.value = ConnectionState.CONNECTED
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                try {
                    val json = JSONObject(text)
                    val event = parseNegotiationEvent(json)
                    _negotiationEvents.tryEmit(event)

                    // Update current negotiation if data is present
                    if (json.has("negotiation")) {
                        val negotiation = parseNegotiation(json.getJSONObject("negotiation"))
                        _currentNegotiation.value = negotiation
                    }
                } catch (e: Exception) {
                    // Handle parse error
                }
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                _connectionState.value = ConnectionState.ERROR
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(1000, null)
                _connectionState.value = ConnectionState.DISCONNECTED
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                _connectionState.value = ConnectionState.DISCONNECTED
            }
        })
    }

    /**
     * Disconnect WebSocket
     */
    fun disconnectWebSocket() {
        webSocket?.close(1000, "User disconnected")
        webSocket = null
        _connectionState.value = ConnectionState.DISCONNECTED
    }

    /**
     * Calculate platform fee based on tiered pricing
     * $1 for fares ≤ $35
     * $2 for fares $35-$70
     * $3 for fares > $70
     */
    fun calculatePlatformFee(fare: Double): Double {
        return when {
            fare <= 35.0 -> 1.0
            fare <= 70.0 -> 2.0
            else -> 3.0
        }
    }

    /**
     * Generate quick offer options (60%, 70%, 80%, 90% of suggested price)
     */
    fun getQuickOfferOptions(suggestedPrice: Double): List<Double> {
        return listOf(0.6, 0.7, 0.8, 0.9).map { percentage ->
            Math.round(suggestedPrice * percentage * 100) / 100.0
        }
    }

    private fun parseNegotiation(json: JSONObject): Negotiation {
        return Negotiation(
            id = json.optString("id", ""),
            orderId = json.optString("order_id", null),
            rideId = json.optString("ride_id", null),
            customerId = json.optString("customer_id", ""),
            driverId = json.optString("driver_id", null),
            suggestedPrice = json.optDouble("suggested_price", 0.0),
            currentPrice = json.optDouble("current_price", 0.0),
            customerOffer = if (json.has("customer_offer")) json.optDouble("customer_offer") else null,
            driverOffer = if (json.has("driver_offer")) json.optDouble("driver_offer") else null,
            platformFee = json.optDouble("platform_fee", 1.0),
            status = NegotiationStatus.valueOf(
                json.optString("status", "PENDING").uppercase()
            ),
            createdAt = json.optLong("created_at", System.currentTimeMillis()),
            expiresAt = json.optLong("expires_at", System.currentTimeMillis() + 300000)
        )
    }

    private fun parseNegotiationEvent(json: JSONObject): NegotiationEvent {
        val typeString = json.optString("type", "PRICE_UPDATE")
        val type = try {
            NegotiationEventType.valueOf(typeString.uppercase())
        } catch (e: Exception) {
            NegotiationEventType.PRICE_UPDATE
        }

        val data = mutableMapOf<String, Any>()
        json.keys().forEach { key ->
            if (key != "type" && key != "negotiation_id") {
                data[key] = json.get(key)
            }
        }

        return NegotiationEvent(
            type = type,
            negotiationId = json.optString("negotiation_id", ""),
            data = data
        )
    }
}

File: apps/ios/eatfair-ios-shared/Sources/EatFairShared/Services/NegotiationService.swift
import Foundation
import os

private let logger = Logger(subsystem: "com.dollorai.shared", category: "NegotiationService")
import Combine

/// Dollor.ai Negotiation Service
/// Real-time price negotiation between independent drivers and customers
/// MATCHMAKING PLATFORM - Platform suggests prices, parties negotiate freely
public class NegotiationService: ObservableObject {
    public static let shared = NegotiationService()

    // MARK: - Configuration
    private var baseURL: String {
        return AppConfig.shared.negotiationServiceURL
    }

    @Published public var isLoading = false
    @Published public var error: String?
    @Published public var activeNegotiation: Negotiation?

    private var cancellables = Set<AnyCancellable>()
    private var webSocket: URLSessionWebSocketTask?

    private init() {}

    // MARK: - Negotiation Models

    public enum NegotiationType: String, Codable {
        case rideshare = "rideshare"
        case foodDelivery = "food_delivery"
    }

    public enum NegotiationStatus: String, Codable {
        case pending = "pending"
        case customerCountered = "customer_countered"
        case driverCountered = "driver_countered"
        case accepted = "accepted"
        case rejected = "rejected"
        case expired = "expired"
        case cancelled = "cancelled"
    }

    public struct Negotiation: Codable, Identifiable {
        public let id: String
        public let negotiationType: NegotiationType
        public let rideId: String?
        public let orderId: String?
        public let customerId: String
        public let driverId: String?
        public let pickupLat: Double
        public let pickupLng: Double
        public let dropoffLat: Double
        public let dropoffLng: Double
        public let distanceMiles: Double
        public let estimatedMinutes: Int
        public let platformSuggestedPrice: Double
        public let currentPrice: Double
        public let customerOffer: Double?
        public let driverAsk: Double?
        public let platformFee: Double
        public let status: NegotiationStatus
        public let expiresAt: String
        public let createdAt: String

        enum CodingKeys: String, CodingKey {
            case id = "negotiation_id"
            case negotiationType = "negotiation_type"
            case rideId = "ride_id"
            case orderId = "order_id"
            case customerId = "customer_id"
            case driverId = "driver_id"
            case pickupLat = "pickup_lat"
            case pickupLng = "pickup_lng"
            case dropoffLat = "dropoff_lat"
            case dropoffLng = "dropoff_lng"
            case distanceMiles = "distance_miles"
            case estimatedMinutes = "estimated_minutes"
            case platformSuggestedPrice = "platform_suggested_price"
            case currentPrice = "current_price"
            case customerOffer = "customer_offer"
            case driverAsk = "driver_ask"
            case platformFee = "platform_fee"
            case status
            case expiresAt = "expires_at"
            case createdAt = "created_at"
        }
    }

    public struct QuickOfferOption: Identifiable {
        public let id = UUID()
        public let percentage: Int
        public let amount: Double
        public let label: String

        public init(percentage: Int, basePrice: Double) {
            self.percentage = percentage
            self.amount = basePrice * Double(percentage) / 100.0
            self.label = "\(percentage)%"
        }
    }

    // MARK: - Create Negotiation

    /// Start a new negotiation for a ride or delivery
    public func createNegotiation(
        type: NegotiationType,
        customerId: String,
        pickupLat: Double,
        pickupLng: Double,
        dropoffLat: Double,
        dropoffLng: Double,
        distanceMiles: Double,
        estimatedMinutes: Int,
        rideId: String? = nil,
        orderId: String? = nil,
        completion: @escaping (Result<Negotiation, Error>) -> Void
    ) {
        guard let url = URL(string: "\(baseURL)/api/negotiations") else {
            completion(.failure(NegotiationError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "negotiation_type": type.rawValue,
            "customer_id": customerId,
            "pickup_lat": pickupLat,
            "pickup_lng": pickupLng,
            "dropoff_lat": dropoffLat,
            "dropoff_lng": dropoffLng,
            "distance_miles": distanceMiles,
            "estimated_minutes": estimatedMinutes,
            "ride_id": rideId as Any,
            "order_id": orderId as Any
        ]

        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        isLoading = true
        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                self?.isLoading = false

                if let error = error {
                    self?.error = error.localizedDescription
                    completion(.failure(error))
                    return
                }

                guard let data = data else {
                    completion(.failure(NegotiationError.noData))
                    return
                }

                do {
                    let negotiation = try JSONDecoder().decode(Negotiation.self, from: data)
                    self?.activeNegotiation = negotiation
                    completion(.success(negotiation))
                } catch {
                    self?.error = "Failed to decode negotiation: \(error.localizedDescription)"
                    completion(.failure(error))
                }
            }
        }.resume()
    }

    // MARK: - Customer Counter Offer

    /// Customer makes a counter offer
    public func customerCounterOffer(
        negotiationId: String,
        customerId: String,
        offerAmount: Double,
        completion: @escaping (Result<Negotiation, Error>) -> Void
    ) {
        guard let url = URL(string: "\(baseURL)/api/negotiations/\(negotiationId)/customer-offer") else {
            completion(.failure(NegotiationError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "customer_id": customerId,
            "offer_amount": offerAmount
        ]

        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        performRequest(request, completion: completion)
    }

    // MARK: - Driver Counter Offer

    /// Driver makes a counter offer
    public func driverCounterOffer(
        negotiationId: String,
        driverId: String,
        askAmount: Double,
        completion: @escaping (Result<Negotiation, Error>) -> Void
    ) {
        guard let url = URL(string: "\(baseURL)/api/negotiations/\(negotiationId)/driver-offer") else {
            completion(.failure(NegotiationError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "driver_id": driverId,
            "ask_amount": askAmount
        ]

        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        performRequest(request, completion: completion)
    }

    // MARK: - Accept Negotiation

    /// Accept the current negotiation price
    public func acceptNegotiation(
        negotiationId: String,
        participantType: String,  // "customer" or "driver"
        participantId: String,
        completion: @escaping (Result<Negotiation, Error>) -> Void
    ) {
        guard let url = URL(string: "\(baseURL)/api/negotiations/\(negotiationId)/accept") else {
            completion(.failure(NegotiationError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "participant_type": participantType,
            "participant_id": participantId
        ]

        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        performRequest(request, completion: completion)
    }

    // MARK: - Quick Offer Options

    /// Get quick offer options (60%, 70%, 80%, 90% of suggested price)
    public func getQuickOfferOptions(suggestedPrice: Double) -> [QuickOfferOption] {
        return [60, 70, 80, 90].map { percentage in
            QuickOfferOption(percentage: percentage, basePrice: suggestedPrice)
        }
    }

    // MARK: - WebSocket Connection

    /// Connect to real-time negotiation updates
    public func connectWebSocket(negotiationId: String, participantType: String, participantId: String) {
        let wsURLString = baseURL.replacingOccurrences(of: "https://", with: "wss://")
            .replacingOccurrences(of: "http://", with: "ws://")
        guard let url = URL(string: "\(wsURLString)/ws/negotiation/\(negotiationId)?participant_type=\(participantType)&participant_id=\(participantId)") else {
            return
        }

        let session = URLSession(configuration: .default)
        webSocket = session.webSocketTask(with: url)
        webSocket?.resume()

        receiveMessage()
    }

    private func receiveMessage() {
        webSocket?.receive { [weak self] result in
            switch result {
            case .success(let message):
                switch message {
                case .string(let text):
                    if let data = text.data(using: .utf8),
                       let negotiation = try? JSONDecoder().decode(Negotiation.self, from: data) {
                        DispatchQueue.main.async {
                            self?.activeNegotiation = negotiation
                        }
                    }
                case .data(let data):
                    if let negotiation = try? JSONDecoder().decode(Negotiation.self, from: data) {
                        DispatchQueue.main.async {
                            self?.activeNegotiation = negotiation
                        }
                    }
                @unknown default:
                    break
                }
                self?.receiveMessage()
            case .failure(let error):
                #if DEBUG
                logger.info("[NegotiationService] WebSocket error: \(error)")
                #endif
            }
        }
    }

    /// Disconnect WebSocket
    public func disconnectWebSocket() {
        webSocket?.cancel(with: .goingAway, reason: nil)
        webSocket = nil
    }

    // MARK: - Helper Methods

    private func performRequest(_ request: URLRequest, completion: @escaping (Result<Negotiation, Error>) -> Void) {
        isLoading = true
        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                self?.isLoading = false

                if let error = error {
                    self?.error = error.localizedDescription
                    completion(.failure(error))
                    return
                }

                guard let data = data else {
                    completion(.failure(NegotiationError.noData))
                    return
                }

                do {
                    let negotiation = try JSONDecoder().decode(Negotiation.self, from: data)
                    self?.activeNegotiation = negotiation
                    completion(.success(negotiation))
                } catch {
                    self?.error = "Failed to decode response: \(error.localizedDescription)"
                    completion(.failure(error))
                }
            }
        }.resume()
    }

    // MARK: - Platform Fee Calculation

    /// Calculate tiered platform fee for rideshare
    public func calculatePlatformFee(fareAmount: Double) -> Double {
        return AppConfig.shared.calculateRidesharePlatformFee(fareAmount: fareAmount)
    }

    /// Get platform fee description
    public func getPlatformFeeDescription(fareAmount: Double) -> String {
        let fee = calculatePlatformFee(fareAmount: fareAmount)
        return String(format: "$%.0f Platform Fee", fee)
    }
}

// MARK: - Error Types

public enum NegotiationError: LocalizedError {
    case invalidURL
    case noData
    case negotiationExpired
    case notAuthorized

    public var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .noData:
            return "No data received"
        case .negotiationExpired:
            return "Negotiation has expired"
        case .notAuthorized:
            return "Not authorized to perform this action"
        }
    }
}
